### PR TITLE
Support I/O and clone for NSG

### PR DESCRIPTION
**Pull request by KinglittleQ**
This PR added IO and clone support to NSG.

### CODE REVIEW

File: faiss/impl/NSG.cpp [hunk 1]
```diff
@@ -285,7 +285,7 @@ void NSG::search_on_graph(
         int pool_size,
         std::vector<Neighbor>& retset,
         std::vector<Node>& fullset) const {
-    RandomGenerator gen(rand());
+    RandomGenerator gen(0x1234);
```
Comment (h-vetinari) [inline faiss/impl/NSG.cpp:288]: Not very random if you always initialize with the same seed, or am I missing something?

Comment (KinglittleQ) [reply faiss/impl/NSG.cpp:288]: `rand()` is not thread-safe and I'm trying to make the process of searching deterministic. This change does not harm the performance according to the experiment I made.

**Comment by KinglittleQ** [comment]
## Updated:
Fix a bug that will occur when search_L = -1.

Before:
```
search_L -1        0.133 ms per query, R@1 0.2889, missing rate 0.0000
```

After:
```
search_L -1        0.005 ms per query, R@1 0.2904, missing rate 0.0000
```

File: faiss/IndexNSG.cpp [hunk 2]
```diff
@@ -85,7 +85,8 @@ void IndexNSG::search(
             storage,
             "Please use IndexNSGFlat (or variants) instead of IndexNSG directly");
 
-    idx_t check_period = InterruptCallback::get_period_hint(d * nsg.search_L);
```
Comment (KinglittleQ) [inline faiss/IndexNSG.cpp:88]: When nsg.search_L = -1, it will cause unexpected behaviour.

**Comment by mdouze** [comment]
Look good!

**Pull request closed without merging**